components/oauth/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java

```diff
@@ -228,89 +228,110 @@ public OAuth2AccessTokenRespDTO issueAccessToken(OAuth2AccessTokenReqDTO tokenRe
-                            org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(
-                                    revokeRequestDTO.getConsumerKey(), refreshTokenDO.getAuthorizedUser(),
-                                    OAuth2Util.buildScopeString(refreshTokenDO.getScope()));
-                            org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(
-                                    revokeRequestDTO.getConsumerKey(), refreshTokenDO.getAuthorizedUser());
-                            org.wso2.carbon.identity.oauth.OAuthUtil.clearOAuthCache(refreshTokenDO.getAccessToken());
-                            tokenMgtDAO.revokeTokens(new String[]{refreshTokenDO.getAccessToken()});
-                            addRevokeResponseHeaders(revokeResponseDTO, refreshTokenDO.getAccessToken(),
-                                    revokeRequestDTO.getToken(), refreshTokenDO.getAuthorizedUser());
+                    accessTokenDO = tokenMgtDAO.retrieveAccessToken(revokeRequestDTO.getToken(), true);
+                    if (accessTokenDO == null) {
+
+                        refreshTokenDO = tokenMgtDAO
+                                .validateRefreshToken(revokeRequestDTO.getConsumerKey(), revokeRequestDTO.getToken());
+
+                        if (refreshTokenDO == null ||
+                                StringUtils.isEmpty(refreshTokenDO.getRefreshTokenState()) ||
+                                !(OAuthConstants.TokenStates.TOKEN_STATE_ACTIVE
+                                        .equals(refreshTokenDO.getRefreshTokenState()) ||
+                                        OAuthConstants.TokenStates.TOKEN_STATE_EXPIRED
+                                                .equals(refreshTokenDO.getRefreshTokenState()))) {
+                            return revokeResponseDTO;
                         }
                     }
                 }
+
+                String grantType = StringUtils.EMPTY;
+
+                if (accessTokenDO != null) {
+                    grantType = accessTokenDO.getGrantType();
+                } else if (refreshTokenDO != null) {
+                    grantType = refreshTokenDO.getGrantType();
+                }
+
+                if (!StringUtils.equals(OAuthConstants.GrantTypes.IMPLICIT, grantType) &&
+                        !OAuth2Util.authenticateClient(revokeRequestDTO.getConsumerKey(),
+                                revokeRequestDTO.getConsumerSecret())) {
+
+                    OAuthRevocationResponseDTO revokeRespDTO = new OAuthRevocationResponseDTO();
+                    revokeRespDTO.setError(true);
+                    revokeRespDTO.setErrorCode(OAuth2ErrorCodes.UNAUTHORIZED_CLIENT);
+                    revokeRespDTO.setErrorMsg("Unauthorized Client");
+
+                    return revokeRespDTO;
+                }
+
+                if (refreshTokenDO != null) {
+
+                    org.wso2.carbon.identity.oauth.OAuthUtil
```

Comment:
import the OAuthUtil class at imports. That will improve the readability of the code